apps/api/src/billing/services/stripe-webhook/stripe-webhook.service.ts

```diff
@@ -104,11 +87,135 @@ export class StripeWebhookService {
+        receiptUrl = charge.receipt_url ?? undefined;
+      } catch (error) {
+        this.logger.warn({
+          event: "CHARGE_DETAILS_FETCH_FAILED",
+          chargeId,
+          error
+        });
+      }
+    }
+
+    await this.stripeTransactionRepository.updateStatusByPaymentIntentId(paymentIntent.id, {
+      status: "succeeded",
+      stripeChargeId: chargeId,
+      paymentMethodType: paymentMethodDetails,
+      cardBrand,
+      cardLast4,
+      receiptUrl
+    });
+
     // Use amount_received when available (for partial captures), otherwise use amount
     const paymentAmount = paymentIntent.amount_received ?? paymentIntent.amount;
     await this.refillService.topUpWallet(paymentAmount, user.id);
   }
 
+  @WithTransaction()
+  async handlePaymentIntentFailed(event: Stripe.PaymentIntentPaymentFailedEvent) {
+    const paymentIntent = event.data.object;
+    const errorMessage = paymentIntent.last_payment_error?.message ?? "Payment failed";
+
+    await this.stripeTransactionRepository.updateStatusByPaymentIntentId(paymentIntent.id, {
+      status: "failed",
+      errorMessage
+    });
+
+    this.logger.warn({
+      event: "PAYMENT_INTENT_FAILED",
+      paymentIntentId: paymentIntent.id,
+      errorMessage
+    });
+  }
+
+  @WithTransaction()
+  async handlePaymentIntentCanceled(event: Stripe.PaymentIntentCanceledEvent) {
+    const paymentIntent = event.data.object;
+
+    await this.stripeTransactionRepository.updateStatusByPaymentIntentId(paymentIntent.id, {
+      status: "canceled"
+    });
+
+    this.logger.info({
+      event: "PAYMENT_INTENT_CANCELED",
+      paymentIntentId: paymentIntent.id
+    });
+  }
+
+  @WithTransaction()
+  async handleChargeRefunded(event: Stripe.ChargeRefundedEvent) {
+    const charge = event.data.object;
+    const customerId = charge.customer as string;
+
+    if (!customerId) {
+      this.logger.error({ event: "CHARGE_REFUNDED_MISSING_CUSTOMER_ID", chargeId: charge.id });
+      return;
+    }
+
+    const refundedAmount = this.calculateRefundDelta(event);
+    if (refundedAmount <= 0) {
+      this.logger.warn({ event: "CHARGE_REFUNDED_NO_DELTA", chargeId: charge.id, totalRefunded: charge.amount_refunded });
+      return;
+    }
+
+    const user = await this.userRepository.findOneBy({ stripeCustomerId: customerId });
+    if (!user) {
+      this.logger.error({ event: "CHARGE_REFUNDED_USER_NOT_FOUND", customerId, chargeId: charge.id });
+      return;
+    }
+
+    const transaction = await this.stripeTransactionRepository.findByChargeId(charge.id);
+    const isFullyRefunded = charge.refunded;
+
+    if (transaction && isFullyRefunded) {
+      await this.stripeTransactionRepository.updateById(transaction.id, { status: "refunded" });
+    } else if (!transaction) {
+      this.logger.warn({ event: "CHARGE_REFUNDED_NO_TRANSACTION", chargeId: charge.id });
+    }
+
+    await this.refillService.reduceWalletBalance(refundedAmount, user.id);
```

Comment:
issue: I think this may backfire. Consider the use case when a transaction failed to sync with blockchain and user never got credits for the stripe transaction. Such tx should only be refunded in stripe. I'm guessing the refund on our side should only happen if there's a related record in successful state in the db.

```suggestion
    if (transaction) await this.refillService.reduceWalletBalance(refundedAmount, user.id);
```